### ISSUE TITLE
SRP0014 and SRD0018 management

### ISSUE THREAD

**Issue by niphlod** [ErikEJ/SqlServer.Rules#712]
When you try to analyze this piece of code

```
DECLARE @body NVARCHAR(MAX) = N'
{
	"data" : [
		{
			"originalId" : "abc1",
			"currentId": "abc2"
		},
		{
			"originalId" : "cde1",
			"currentId": "cde2"
		}
	]
}
';
DECLARE @someTable TABLE (
	originalId  VARCHAR(50),
	currentId   VARCHAR(50)
);

UPDATE dst
SET currentId = bodyItems.currentId
FROM @someTable dst
INNER JOIN OPENJSON(@body, '$.data') WITH (
	originalId  VARCHAR(50),
	currentId   VARCHAR(50)
) bodyItems
ON bodyItems.originalId = dst.originalId;
```

both SRP0014 and SRD0018 are raised, IMHO improperly.

**Comment by ErikEJ** [comment]
@niphlod Could you elaborate why? (helps me fix the potential false positive)

**Comment by niphlod** [comment]
hm, okay.
SRD0018: UPDATE statements not having a WHERE clause .... I'm restricting it via an INNER JOIN, so a "filter" is there . 
Given I need to "project" the value I need to update, having a "filter" via a "similar" logic would be to use an EXISTS, but it'll force to do subqueries for the update itself, so .... is there another way ? 
Maybe a valid warning would be for a JOIN that matches more than one record, but it *might* be out of the scope of a static analyzer.

SRP0014: Avoid the use of table variables in join clauses.... but if you look at the documentation is a bit more "expanded" 
"""
Avoid the use of table variables in join clauses planning and maintenace hazard.
"""
so, I'm assuming "planning" is referred to SQL execution plan and "maintena*N*ce" is ... on the user ? Dunno, but, really, don't care.

In the extremely contrived example, I'm updating a table variable IN-PLACE. there's nothing that could *avoid* me using a table variable in join ^_^ .

Then, let's assume a more-expanded example .... That I'm targeting a "real" table.

as in e.g.
```
CREATE TABLE dbo.test (
	originalId  VARCHAR(50),
	currentId   VARCHAR(50)
);

UPDATE dst
SET currentId = bodyItems.currentId
FROM dbo.test dst
INNER JOIN OPENJSON(@body, '$.data') WITH (
	originalId  VARCHAR(50),
	currentId   VARCHAR(50)
) bodyItems
ON bodyItems.originalId = dst.originalId;
```
Fine, SRD0018 doesn't pop up.

But, it pops up for this
```
DECLARE @someTable TABLE (
	originalId  VARCHAR(50),
	currentId   VARCHAR(50)
);

UPDATE dst
SET currentId = bodyItems.currentId
FROM dbo.test dst
INNER JOIN @someTable bodyItems
ON bodyItems.originalId = dst.originalId;
```

Now, I'm not sure if the cardinality estimator "estimates" OPENJSON "better" than table variables.

Is it true that using a temp table, with "real" statistics, is usually faster than a table variable which, depending on SQL server version, gets its own "probably skewed" cardinality estimator ? Yeah, but only in cases where there are at least hundreds of rows.
So, unless it's a "hint" (read "*try* to avoid the use of table variables...if there are more than 100 records") it doesn't apply "blanket style".
Also, what's the balance between "creating temp, filling it with data, joining with temp" and "joining with table variable" ? It's a case-by-case based, and probably would need cardinality info and execution statistics to know which approach is better, leading to an "out-of-scope" env for a static analyzer.

Another "funny" thing (related to SRP0014, but unrelated to SRD0018) is that introducing a temp table creation
```
CREATE TABLE #test (
	originalId  VARCHAR(50),
	currentId   VARCHAR(50)
);
UPDATE dst
SET currentId = bodyItems.currentId
FROM dbo.test dst
INNER JOIN #test bodyItems
ON bodyItems.originalId = dst.originalId;
```
removes all warnings, SRP0014 included.

And that leads to a different "bug" alltoghether .... how to be sure that the analyzer is working when no warnings are raised ?
Appending 
```
WHERE bodyItems.originalId <> ''
```
to the update doesn't trigger SRP0006 (which has its own debatable "merit", but at least is easy to "produce").
My *assumption* is that temp table creation "broke" the analyzer.

**Comment by ErikEJ** [comment]
@niphlod OK, thanks for the additional information going in many directions.

I will try to create targeted issues for the things you are bringing up here....

**Comment by ErikEJ** [comment]
@niphlod 

```sql
CREATE TABLE #test (
	originalId  VARCHAR(50),
	currentId   VARCHAR(50)
);
UPDATE dst
SET currentId = bodyItems.currentId
FROM dbo.test dst
INNER JOIN #test bodyItems
ON bodyItems.originalId = dst.originalId;
```

SRP0014 is "Avoid the use of table variables in join clauses." https://github.com/ErikEJ/SqlServer.Rules/blob/master/docs/Performance/SRP0014.md

Why would you expect it to be triggered in this case?

**Comment by ErikEJ** [comment]
@niphlod I have created som targeted issues, and will close this - feel free to follow up on the focused issues

**Comment by niphlod** [comment]
> [@niphlod](https://github.com/niphlod)
> 
> CREATE TABLE #test (
> 	originalId  VARCHAR(50),
> 	currentId   VARCHAR(50)
> );
> UPDATE dst
> SET currentId = bodyItems.currentId
> FROM dbo.test dst
> INNER JOIN #test bodyItems
> ON bodyItems.originalId = dst.originalId;
> SRP0014 is "Avoid the use of table variables in join clauses." https://github.com/ErikEJ/SqlServer.Rules/blob/master/docs/Performance/SRP0014.md
> 
> Why would you expect it to be triggered in this case?

@ErikEJ 
I don't want SRP0014 to be triggered... was rather pointing to the fact that SRD0018 isn't triggered, another "pointer" to *maybe* the inability to parse (and raise warnings) when temp tables are created.
